lighthouse/helpers/plates.py

```diff
@@ -529,116 +487,216 @@ def create_cherrypicked_post_body(
     return {"data": {"type": "plates", "attributes": body}}
 
 
-def sample_subjects(samples):
+# WARN - on refactoring, be careful not to lose the distributed functionality where
+# None or empty samples returns None
+def get_source_plates_for_samples(samples):
+    barcodes = get_unique_plate_barcodes(samples)
+    return find_source_plates(query_for_source_plate_uuids(barcodes))
+
+
+def find_source_plates(query: Dict[str, Any]) -> Optional[List[Dict[str, Any]]]:
+    if query is None:
+        return None
+
+    source_plates = app.data.driver.db.source_plates
+
+    source_plate_documents = list(source_plates.find(query))
+
+    logger.info(f"Found {len(source_plate_documents)} source plates")
+
+    return source_plate_documents
+
+
+def get_unique_plate_barcodes(samples):
+    barcodes = set()
+    for sample in samples:
+        barcodes.add(sample[FIELD_PLATE_BARCODE])
+    return list(barcodes)
+
+
+def query_for_source_plate_uuids(barcodes):
+    if barcodes is None or (len(barcodes) == 0):
+        return None
+
+    return {"$or": [{FIELD_BARCODE: barcode} for barcode in barcodes]}
+
+
+def construct_cherrypicking_plate_failed_message(
+    barcode: str, user_id: str, robot_serial_number: str, failure_type: str
+) -> Tuple[List[str], Optional[Message]]:
+    try:
+        subjects, errors = [], []
+
+        # Add robot and destination plate subjects
+        subjects.append(__robot_subject(robot_serial_number))
+        subjects.append(construct_destination_plate_message_subject(barcode))
+
+        # Try to add sample and source plate subjects
+        dart_samples = None
+        try:
+            dart_samples = find_dart_source_samples_rows(barcode)
+        except Exception:
+            # a failed DART connection is valid:
+            # it may be caused by the failure the user is trying to record
```

Comment:
I don't understand this...

Reply:
So, yeah, this and the still sending a message if no samples can be found in DART is how we discussed we'd handle the failure. E.g. the failure might have caused the DART DB to fall over. If so, we still want to report a plate failing, but might not be able to get samples information for that plate. This is the other side of the coin to the requirement _"If we know what samples are being 'lost' in the failed plate, these should be recorded in an event"_: we are unable to determine which samples were 'lost', so we can't record them

Reply:
OK, thanks.